Update user delete response

peiyaoh opened the pull request:
### Description

- Update user delete return value to be boolean to align with data manager.


### Related Jira Tickets

- https://umich-mia.atlassian.net/browse/JST-97?atlOrigin=eyJpIjoiMThkMzhjNzBjNWQ3NGYwNzkyYWExYmZkODA5NjgwZDciLCJwIjoiaiJ9


### Comments 

-


Please ensure all items are complete before requesting a review:

- [ ] **Code Completed**: All code for the feature or bugfix has been implemented.
- [ ] **Unit Tests Added**: Unit tests are written for all new functionality, covering edge cases where applicable.
- [ ] **Integration Tests (where necessary)**: Integration tests are added for interactions between modules or components.
- [ ] **All Tests Passing**: All unit and integration tests are passing.
- [ ] **Type Safety Checked**: Ensure TypeScript types are correctly defined and used. Avoid `any` types without proper justification.
- [ ] **Typedoc Comments Added**: Ensure all functions, classes, interfaces, and types are documented using TypeDoc syntax.
- [ ] **Code Formatting**: All code files are formatted according to the project’s standards (e.g., Prettier, ESLint) and no linting errors remain.
- [ ] **No Console Logs**: Remove any `console.log` or other debugging output from the code.
- [ ] **Documentation Updated**: Update relevant project documentation (e.g., README, configuration guides) if necessary.

---

peiyaoh committed:
Return boolean from user deletion methods

Changed user deletion methods to return a boolean indicating success instead of void. Updated related tests and method signatures in JustInWrapper and UserManager to reflect this change, improving clarity on operation outcomes.

peiyaoh committed:
Add debug log for logWarnStub arguments in test

Added a console.log statement to print the arguments passed to logWarnStub in the user-manager unit test. This helps with debugging test behavior when a user already exists.

peiyaoh committed:
Comment out selected UserManager unit tests

Several test suites and test cases in user-manager.unit.test.ts have been commented out, likely for debugging or to temporarily disable them. No production code is affected.

peiyaoh committed:
Uncomment and update UserManager unit tests

Restores previously commented-out test suites for UserManager and updates stubs and assertions to reflect current implementation. Ensures delete methods return correct values and improves test coverage.

peiyaoh committed:
Reorder assertions in deleteUserById test

Moved the assertion for the result of deleteUserById to follow the stub call assertion for improved test clarity.

peiyaoh committed:
Clean up user-manager unit tests

Removed unnecessary console.log and extra blank lines from user-manager.unit.test.ts. Combined variable declarations for improved readability.

jadrubio commented on src/user-manager/user-manager.ts:
```diff
@@ -316,10 +316,11 @@ const updateUserById = async (
   _checkInitialization();
-  await dm.removeItemFromCollection(USERS, userId);
+  const result = await dm.removeItemFromCollection(USERS, userId);
   _users.delete(userId);
```
We should do a check to see if the result is true before removing it from _users

peiyaoh replied:
You are right. Let me add it.

peiyaoh replied:
@jadrubio Just add the check. Please take a look.

One interesting aspect of defining behavior (in MongoDB) is that the `deleteOne` method will return `acknowledged == true` even if no item matched the deletion criteria. In a sense, it behaves as if the item was deleted. I am trying to replicate this behavior in the UserManager as well.

jadrubio commented on src/user-manager/user-manager.ts:
```diff
@@ -329,10 +330,11 @@ const deleteUserById = async (userId: string): Promise<void> => {
   const theUser: JUser = await getUserByUniqueIdentifier(uniqueIdentifier) as JUser;
-  await deleteUserById(theUser.id);
+  const result = await deleteUserById(theUser.id);
   _users.delete(theUser.id);
```
We should do a check to see if the result is true before removing it from _users

peiyaoh replied:
You are right. Let me add it.

peiyaoh replied:
@jadrubio Just add the check. Please take a look.

peiyaoh committed:
Fix cache deletion on failed user removal

Update deleteUserById and deleteUserByUniqueIdentifier to only remove users from the cache if the database deletion is successful. Added corresponding unit tests to verify correct cache behavior when deletion fails.

peiyaoh committed:
Update user deletion to return success status

Changed deleteUserById and deleteUserByUniqueIdentifier to return a boolean indicating if deletion was successful, instead of void. Updated related test to match new behavior and improve stub argument matching.

peiyaoh committed:
Add test for cache retention on failed user deletion

Introduces a unit test to verify that a user is not removed from the in-memory cache if the database deletion operation fails in UserManager.

peiyaoh committed:
Merge pull request #50 from MIACollaborative/check-delete-cache

Check delete cache

numerator commented on src/user-manager/user-manager.ts:
```diff
@@ -314,25 +314,28 @@ const updateUserById = async (
+  const theUser: JUser | null = await getUserByUniqueIdentifier(uniqueIdentifier);
+  const userId = theUser?.id as any;
+  const result = await deleteUserById(userId);
+  if(result) _users.delete(userId);
```
Doesn't `deleteUserById()` already delete from the cache?

peiyaoh replied:
Good point. I just removed cache deletion from this function.

numerator commented on src/user-manager/user-manager.ts:
```diff
@@ -314,25 +314,28 @@ const updateUserById = async (
   _checkInitialization();
-  await dm.removeItemFromCollection(USERS, userId);
-  _users.delete(userId);
+  const result = await dm.removeItemFromCollection(USERS, userId);
```
digging into this, I see that this is the code for `removeItemFromCollection` in mongo-data-manager:
```
const removeItemFromCollection = async (
  collectionName: string,
  id: string
): Promise<boolean> => {
  ensureInitialized();
  const objectId = toObjectId(id);
  if (!objectId) return false;

  try {
    const { acknowledged } = await _db!
      .collection(collectionName)
      .deleteOne({ _id: objectId });
    return acknowledged;
  } catch (error) {
    return handleDbError(
      `Error removing item with id ${id} from ${collectionName}`,
      error
    );
  }
};
```

Are we sure that "acknowledged" will be false if the doc is not actually removed? Like if an objectId is provided but that document doesn't exist? I think the result returned from `deleteOne()` in such a case will be `{acknowledged: true, deletedCount: 0}`. It would be good to verify how this works and, if I'm right, what should we do?

peiyaoh replied:
Based on my read of the MongoDB documentation, acknowledged == true means the delete option has been attempted. The deletedCount represents the actual number of documents deleted (provided that "acknowledged" is true).

I think we can stay with the current definition, and add a warning when deletedCount == 0 (document not found, effectively complete the delete) so that devs are aware.

This follows the same model in updateItem, when the update operation is attempted that does not change any values (effectively complete the update), we log the warning and still return the document.

Alternative 1: `return deletedCount > 0;` to indicate whether an item has been deleted.

Alternative 2: we could still return the document (if found) || null (not found), so that devs can have more information. Mongo provides findOneAndDelete() for this. It does require a write lock, which could impact performance.
https://www.mongodb.com/docs/manual/reference/method/db.collection.findOneAndDelete/


Terminology

Acknowledged: A boolean acknowledged as true if the operation ran with write concern or false if write concern was disabled.
https://www.mongodb.com/docs/manual/reference/method/db.collection.deleteOne/

Write concern
Specifies whether a write operation has succeeded. Write concern allows your application to detect insertion errors or unavailable MongoDB instances. 
https://www.mongodb.com/docs/manual/reference/glossary/#std-term-write-concern

Write concern describes the level of acknowledgment requested from MongoDB for write operations to a standalone mongod, replica sets, or sharded clusters.
https://www.mongodb.com/docs/manual/reference/write-concern/

peiyaoh committed:
Log warning when item deletion fails in Mongo manager

Adds a warning log if no item is deleted when attempting to remove an item from a MongoDB collection, helping to identify cases where the item does not exist.

peiyaoh committed:
Merge pull request #53 from MIACollaborative/add-remove-warning

Log warning when item deletion fails in Mongo manager

peiyaoh committed:
Fix indentation in removeItemFromCollection warning

Corrected the indentation of the Log.warn statement in the removeItemFromCollection function for better code readability.

peiyaoh committed:
Return true if user not found in deleteUserByUniqueIdentifier

Added a check to return true when no user is found by unique identifier in deleteUserByUniqueIdentifier. This prevents unnecessary deletion attempts and ensures consistent return values.

peiyaoh committed:
Fix test assertion for user deletion by UID

Corrects the expectation in the test to ensure that removeItemFromCollectionStub is not called when attempting to delete a non-existent user by unique identifier.

peiyaoh committed:
Fix user deletion logic for non-existent users

Corrects the behavior of deleteUserByUniqueIdentifier to return true when attempting to delete a non-existent user. Updates the related unit test to expect the correct stub call.

peiyaoh committed:
Simplify deleteUserByUniqueIdentifier logic

Refactored deleteUserByUniqueIdentifier to remove unnecessary try-catch and redundant checks. The function now directly deletes the user and updates the _users map if successful.

peiyaoh committed:
Update .gitignore and format package.json

Added Yarn-related files to .gitignore for better repository hygiene. Reformatted the keywords and publishConfig fields in package.json for improved readability.

peiyaoh commented:
Hi all,

My digging revealed a nuance in our current implementation regarding the mongo-data-manager deletion logic. I want to highlight the current behavior of this PR.

🔍 Current Deletion Behavior
The mongo-data-manager has two different outcomes for unsuccessful deletions:
- Invalid ID: An invalid ID throws an error at the manager level and returns false. (This signals a technical failure).
- No Match Found: If the ID is valid but no item exists, it returns true (Success - the item is gone).

PR Behavior: This PR, by propagating the result up to user-manager.deleteUserByUniqueIdentifier(), maintains this logic.

⚠️ Impact of Current Logic (Double Deletion)
If a deletion of the same existing user is attempted twice:
- The first attempt will return true (Successful deletion).
- The second attempt will return false (Unsuccessful—no matching item found, but due to the inherited logic path from the called method as deleteUserByUniqueIdentifier can no longer find a matching user with a valid id).

Good News: In this scenario, the second attempt logs an error but does not crash.

💡 Next Steps
I suggest we leave any consideration of changing this behavior until after the core + engine refactoring is complete.

Let me know if you have any questions or concerns about merging this PR with its current behavior.

peiyaoh committed:
Add TODOs to review removeItemFromCollection return value

Added comments to highlight uncertainty about the return value of removeItemFromCollection, specifically regarding cases where objectId is invalid or no items are deleted. These TODOs indicate areas for future review and possible logic changes.
